tests/test_models.py

```diff
@@ -73,6 +73,35 @@ def test_local_gain_control_dict(self, basic_stim):
             assert torch.norm(y[key] - y_hat[key]) < 1e-5
 
 
+class TestLaplacianPyramid(object):
+
+    def test_grad(self, basic_stim):
+        lpyr = po.simul.LaplacianPyramid().to(DEVICE)
+        y = lpyr.analysis(basic_stim)
+        assert y[0].requires_grad
+
+    @pytest.mark.parametrize("n_scales", [3, 4, 5, 6])
+    def test_synthesis(self, curie_img, n_scales):
+        img = curie_img[0:253, 0:234]  # Original 256x256 shape is not good for testing padding
+        lpyr = po.simul.LaplacianPyramid(n_scales=n_scales).to(DEVICE)
+        y = lpyr.analysis(img)
+        img_recon = lpyr.synthesis(y)
+        assert torch.allclose(img, img_recon)
+
+    @pytest.mark.parametrize("n_scales", [3, 4, 5, 6])
+    def test_match_pyrtools(self, curie_img, n_scales):
+        img = curie_img[0:253, 0:234]
+        lpyr_po = po.simul.LaplacianPyramid(n_scales=n_scales).to(DEVICE)
+        y_po = lpyr_po(img)
+        lpyr_pt = pt.pyramids.LaplacianPyramid(img.squeeze().cpu(), height=n_scales)
+        y_pt = [lpyr_pt.pyr_coeffs[(i, 0)] for i in range(n_scales)]
+        assert len(y_po) == len(y_pt)
+        for x_po, x_pt in zip(y_po, y_pt):
+            x_po = x_po.squeeze().detach().cpu().numpy()
+            assert np.abs(x_po - x_pt)[:-2, :-2].max() < 1e-5
+            # TODO: There is some problem with padding on right and bottom edge, should figure out why
```

Comment:
if this is still a problem, make an issue for it

Reply:
Yeah I haven't figured it out yet, will make an issue.

Reply:
I have found the difference in boundary handling. The pyrtools implementation `pt.upConv` performs padding after upsampling. My implementation here `po.tools.upsample_convolve` performs padding before upsampling, and, depending on the parity of the image, sometimes performs additional zero padding after upsampling up to one row/column. The implementation of pyrtools sometimes produces inconsistency: for example, it sometimes effectively performs "reflect2" padding when asked to perform "reflect1" padding due to zeros introduced in upsampling. I have found a way to let my code match pyrtools, but I think my current implementation is better, so I will not change it. I will replace this TODO message in the test file with an explanatory comment.

Reply:
Gotcha, yes please make those changes. It sounds like this is a pyrtools bug then? can you create an issue there?

Reply:
I think it's not necessarily a bug in prytools, it's just a slightly different boundary handling logic and doesn't make a meaningful difference.